### DISCUSSION TITLE
Type can be unknown when using zod optional().catch(undefined)

### DISCUSSION THREAD

**Discussion by popoleeMaster** [colinhacks/zod#2044] [Q&A] [answered]
I have the following GQL interface

```ts
export type ProductFilterInput = {
  readonly duplicated?: InputMaybe<Scalars['Boolean']>;
};
```

which basically mean `duplicated` can be `boolean | undefined | null`

I setup zod to do this check

`duplicated: z.boolean().optional().catch(undefined),`

The issue is this code now throw me 

```
Types of property 'duplicated' are incompatible.
          Type 'unknown' is not assignable to type 'InputMaybe<boolean> | undefined'.
```

Where is this unknown coming from ?

**Comment by JacobWeisenburger** [comment]
Everything seems to be working as expected. Is there something I missed?

```ts
type ProductFilterInput = {
    readonly duplicated?: boolean | null | undefined
}

const schema = z.object( {
    duplicated: z.boolean().nullish().catch( undefined ),
} )
type Data = z.infer<typeof schema>
// type Data = {
//     duplicated?: boolean | null | undefined
// }

console.log( schema.parse( { duplicated: 'foo' } ) )
// { duplicated: undefined }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by popoleeMaster** [reply]
Thanks for the reply 

I have the same issue with your code 
![Screenshot 2023-02-17 at 10 38 51](https://user-images.githubusercontent.com/46583218/219527576-90f4c478-f74b-42ab-b48e-51d757866b39.png)


I am using 3.20.6

https://stackblitz.com/edit/typescript-jp3yib?file=index.ts

maybe a bug ? There should not be unknown in this type at all.

**Reply by popoleeMaster** [reply]
This will throw an error. 

```
type ProductFilterInput = {
  readonly duplicated?: boolean | null | undefined;
};

const schema: z.ZodType<ProductFilterInput>   = z.object({
  duplicated: z.boolean().nullish().catch(undefined),
});
type Data = z.infer<typeof schema>;
```


Reason why this is problematic is, I have some Graphql Schema, and I am using zod like above code to tell, this is my GQL schema, so I want to keep the sync between schema of `z.object`and my GQL schema. (so that id some new field are added or changed in graphql, I would get notified that I am not handling it in zod therefor I could update my zod schema.

Right now with this behavior that wasn't there in some patch ago, I need to create a schema from zod and manually keep the sync with graphql.

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
type ProductFilterInput = {
    readonly duplicated?: boolean | null | undefined
}

const schema = z.object( {
    duplicated: z.boolean().nullish().catch( undefined ),
} ) satisfies z.ZodType<ProductFilterInput, z.ZodTypeDef, unknown>

type Data = z.infer<typeof schema>
// type Data = {
//     duplicated?: boolean | null | undefined
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
